File: backend/src/invitation-policy.ts
export type InvitationState = "SENT" | "ACCEPTED" | "LATER" | "DECLINED" | "EXPIRED" | "CANCELLED" | "COMPLETED";

const transitions: Record<InvitationState, readonly InvitationState[]> = {
  SENT: ["ACCEPTED", "LATER", "DECLINED", "EXPIRED", "CANCELLED"],
  ACCEPTED: ["COMPLETED", "CANCELLED"],
  LATER: ["ACCEPTED", "DECLINED", "EXPIRED"],
  DECLINED: [], EXPIRED: [], CANCELLED: [], COMPLETED: [],
};

export const canTransition = (from: InvitationState, to: InvitationState) => transitions[from].includes(to);

File: app/src/main/java/com/immrtldragon/detoxspace/domain/InvitationPolicy.kt
package com.immrtldragon.detoxspace.domain

object InvitationPolicy {
    private val allowed = mapOf(
        InvitationState.SENDING to setOf(InvitationState.SENT, InvitationState.CANCELLED),
        InvitationState.SENT to setOf(
            InvitationState.ACCEPTED, InvitationState.LATER, InvitationState.DECLINED,
            InvitationState.EXPIRED, InvitationState.CANCELLED,
        ),
        InvitationState.ACCEPTED to setOf(InvitationState.COMPLETED, InvitationState.CANCELLED),
        InvitationState.LATER to setOf(InvitationState.ACCEPTED, InvitationState.DECLINED, InvitationState.EXPIRED),
    )

    fun canTransition(from: InvitationState, to: InvitationState): Boolean =
        to in (allowed[from] ?: emptySet())
}
